exonum-java-binding-common/src/main/java/com/exonum/binding/common/proofs/list/ListProofRootHashCalculator.java

```diff
@@ -0,0 +1,137 @@
+import com.exonum.binding.common.serialization.Serializer;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.TreeMap;
+
+/**
+ * List proof root hash calculator.
+ *
+ * @param <E> the type of elements in the corresponding list
+ */
+public final class ListProofRootHashCalculator<E> implements ListProofVisitor {
+
+  private final CheckingSerializerDecorator<E> serializer;
+
+  private final HashFunction hashFunction;
+
+  private final NavigableMap<Long, E> elements;
+
+  private long index;
+
+  private HashCode calculatedRootHash;
+
+  /**
+   * Creates a new ListProofRootHashCalculator.
+   *
+   * @param serializer a serializer of list elements
+   */
+  public ListProofRootHashCalculator(Serializer<E> serializer) {
+    this(serializer, Hashing.defaultHashFunction());
+  }
+
+  // to easily inject a mock of a calculatedRootHash function.
+  @VisibleForTesting
+  ListProofRootHashCalculator(Serializer<E> serializer, HashFunction hashFunction) {
+    this.serializer = CheckingSerializerDecorator.from(serializer);
+    this.hashFunction = checkNotNull(hashFunction);
+    elements = new TreeMap<>();
+    index = 0;
+    calculatedRootHash = null;
+  }
+
+  @Override
+  public void visit(ListProofBranch branch) {
+    long branchIndex = index;
+    HashCode leftHash = visitLeft(branch, branchIndex);
+    Optional<HashCode> rightHash = visitRight(branch, branchIndex);
+
+    calculatedRootHash = hashFunction.newHasher()
+        .putObject(leftHash, hashCodeFunnel())
+        .putObject(rightHash, (Optional<HashCode> from, PrimitiveSink into) ->
+            from.ifPresent((hash) -> hashCodeFunnel().funnel(hash, into)))
+        .hash();
+  }
+
+  @Override
+  public void visit(ListProofHashNode listProofHashNode) {
+    this.calculatedRootHash = listProofHashNode.getHash();
+  }
+
+  @Override
+  public void visit(ListProofElement value) {
+    assert !elements.containsKey(index) :
+        "Error: already an element by such index in the map: i=" + index
+            + ", e=" + elements.get(index);
+
+    E element = serializer.fromBytes(value.getElement());
+    elements.put(index, element);
+    calculatedRootHash = hashFunction.hashObject(value, ListProofElement.funnel());
+  }
+
+  private HashCode visitLeft(ListProofBranch branch, long branchIndex) {
+    index = 2 * branchIndex;
+    branch.getLeft().accept(this);
+    return calculatedRootHash;
+  }
+
+  private Optional<HashCode> visitRight(ListProofBranch branch, long branchIndex) {
+    index = 2 * branchIndex + 1;
+    calculatedRootHash = null;
+    branch.getRight().ifPresent((right) -> right.accept(this));
+    return Optional.ofNullable(calculatedRootHash);
+  }
+
+  /**
+   * Returns a non-empty collection of list entries: index-element pairs, ordered by indices.
+   *
+   * @throws IllegalStateException if proof is not valid
```

Comment:
I don't see where `IllegalStateException` is thrown
And is it always non-empty?

Reply:
Correct , javadoc should be updated.